Sources/BraintreeCoreSwift/BTAnalyticsService.swift

```diff
@@ -0,0 +1,175 @@
+import Foundation
+
+class BTAnalyticsService: Equatable {
+
+    // MARK: - Internal Properties
+
+    /// A serial dispatch queue that synchronizes access to `analyticsSessions`
+    let sessionsQueue: DispatchQueue = DispatchQueue(label: "com.braintreepayments.BTAnalyticsService")
+
+    /// The HTTP client for communication with the analytics service endpoint. Exposed for testing.
+    var http: BTHTTP?
+
+    /// Defaults to 1, can be overridden
+    var flushThreshold: Int
+
+    /// Dictionary of analytics sessions, keyed by session ID. The analytics service requires that batched events
+    /// are sent from only one session. In practice, BTAPIClient.metadata.sessionID should never change, so this
+    /// is defensive.
+    var analyticsSessions: [String: BTAnalyticsSession] = [:]
+
+    private let apiClient: BTAPIClient
+
+    // MARK: - Initializer
+
+    init(apiClient: BTAPIClient, flushThreshold: Int = 1) {
+        self.apiClient = apiClient
+        self.flushThreshold = flushThreshold
+    }
+
+    // MARK: - Internal Methods
+
+    ///  Tracks an event.
+    ///
+    ///  Events are queued and sent in batches to the analytics service, based on the status of the app and
+    ///  the number of queued events. After exiting this method, there is no guarantee that the event has been sent.
+    /// - Parameter eventName: String representing the event name
+    func sendAnalyticsEvent(_ eventName: String) {
+        DispatchQueue.main.async {
+            self.enqueueEvent(eventName)
+            self.flushIfAtThreshold()
+        }
+    }
+
+    func sendAnalyticsEvent(_ eventName: String, completion: @escaping (Error?) -> Void = { _ in }) {
+        DispatchQueue.main.async {
+            self.enqueueEvent(eventName)
+            self.flush(completion)
+        }
+    }
+
+    func flush(_ completion: @escaping (Error?) -> Void = { _ in }) {
+        apiClient.fetchOrReturnRemoteConfiguration { configuration, error in
+            guard let configuration, error == nil else {
+                if let error {
+                    completion(error)
+                }
+                return
```

Comment:
If both `configuration` and `error` are nil, the completion wouldn't be called. I don't think we have that case in `APIClient.fetchOrReturnRemoteConfiguration()` though. Is this something we want to worry about/consider?

Reply:
We can remove the `error == nil` portion and create an error for if error is nil if we want to guard against this case. I don't _think_ it's likely but better safe than sorry. Thoughts?

Reply:
Actually, running the tests and looking through the rest of the code it doesn't look like we would ever run up against this case as in obj-c we always do when calling `fetchOrReturnRemoteConfiguration`:
```
        if (error) {
            completionBlock(nil, error);
            return;
        }
```
So we should always have either an error or a config (and never a case where both are nil) so I think we are good here.